File: dsa-problems/Recursion/solutions/python/REC-006-subset-sum-exact-count.py
def find_subset(arr: list[int], k: int, target: int) -> list[int]:
    n = len(arr)
    result = []

    def backtrack(index, count, current_sum):
        if count == k:
            return current_sum == target
        
        if index == n:
            return False
        
        # Pruning
        if n - index < k - count:
            return False

        # Option 1: Include
        result.append(arr[index])
        if backtrack(index + 1, count + 1, current_sum + arr[index]):
            return True
        result.pop()

        # Option 2: Exclude
        if backtrack(index + 1, count, current_sum):
            return True
            
        return False

    if backtrack(0, 0, 0):
        return result
    return []


def main():
    import sys
    lines = sys.stdin.read().strip().split('\n')
    if not lines:
        return

    first_line = lines[0].split()
    n = int(first_line[0])
    k = int(first_line[1])
    target = int(first_line[2])

    arr = list(map(int, lines[1].split()))

    result = find_subset(arr, k, target)
    if result:
        print(' '.join(map(str, result)))
    else:
        print("NONE")

if __name__ == "__main__":
    main()

File: dsa-problems/Recursion/solutions/javascript/REC-006-subset-sum-exact-count.js
const readline = require('readline');
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
let tokens = [];
rl.on('line', (line) => { tokens.push(...line.trim().split(/\s+/)); });
rl.on('close', () => {
    if(tokens.length===0) return;
    let ptr = 0;
    
    if(ptr >= tokens.length) return;
    const n = parseInt(tokens[ptr++]);
    const k = parseInt(tokens[ptr++]);
    const target = parseInt(tokens[ptr++]);
    
    const arr = [];
    for(let i=0; i<n; i++) {
        if(ptr < tokens.length) arr.push(parseInt(tokens[ptr++]));
    }
    
    const sol = new Solution();
    const res = sol.findSubset(arr, k, target);
    
    if(res.length === 0) {
        console.log("NONE");
    } else {
        console.log(res.join(' '));
    }
});

class Solution {
    findSubset(arr, k, target) {
        const current = [];
        if (this.backtrack(0, 0, 0, arr, k, target, current)) {
            return current;
        }
        return [];
    }

    backtrack(index, count, currentSum, arr, k, target, current) {
        if (count === k) {
            return currentSum === target;
        }
        if (index === arr.length) {
            return false;
        }
        
        // Pruning
        if (arr.length - index < k - count) {
            return false;
        }

        // Option 1: Include
        current.push(arr[index]);
        if (this.backtrack(index + 1, count + 1, currentSum + arr[index], arr, k, target, current)) {
            return true;
        }
        current.pop();

        // Option 2: Exclude
        if (this.backtrack(index + 1, count, currentSum, arr, k, target, current)) {
            return true;
        }

        return false;
    }
}
